allennlp/modules/bimpm_matching.py

```diff
@@ -0,0 +1,396 @@
+    """
+    Calculate multi-perspective cosine matching between time-steps of vectors
+    of the same length.
+
+    Parameters
+    ----------
+    vector1 : ``torch.Tensor``
+        A tensor of shape ``(batch, seq_len, hidden_size)``
+    vector2 : ``torch.Tensor``
+        A tensor of shape ``(batch, seq_len or 1, hidden_size)``
+    weight : ``torch.Tensor``
+        A tensor of shape ``(num_perspective, hidden_size)``
+
+    Returns
+    -------
+    A tuple of two tensors consisting multi-perspective matching results.
+    The first one is of the shape (batch, seq_len, 1), the second one is of shape
+    (batch, seq_len, num_perspective)
+    """
+    assert vector1.size(0) == vector2.size(0)
+    assert weight.size(1) == vector1.size(2) == vector1.size(2)
+
+    # (batch, seq_len, 1)
+    mv_one = F.cosine_similarity(vector1, vector2, 2).unsqueeze(2)
+
+    # (1, 1, num_perspective, hidden_size)
+    weight = weight.unsqueeze(0).unsqueeze(0)
+
+    # (batch, seq_len, num_perspective, hidden_size)
+    vector1 = weight * vector1.unsqueeze(2)
+    vector2 = weight * vector2.unsqueeze(2)
+
+    mv_many = F.cosine_similarity(vector1, vector2, dim=3)
+
+    return mv_one, mv_many
+
+
+def mpm_pairwise(vector1: torch.Tensor,
+                 vector2: torch.Tensor,
+                 weight: torch.Tensor,
+                 eps: float = 1e-8) -> torch.Tensor:
+    """
+    Calculate multi-perspective cosine matching between each time step of
+    one vector and each time step of another vector.
+
+    Parameters
+    ----------
+    vector1 : ``torch.Tensor``
+        A tensor of shape ``(batch, seq_len1, hidden_size)``
+    vector2 : ``torch.Tensor``
+        A tensor of shape ``(batch, seq_len2, hidden_size)``
+    weight : ``torch.Tensor``
+        A tensor of shape ``(num_perspective, hidden_size)``
+    eps : ``float`` optional, (default = 1e-8)
+        A small value to avoid zero division problem
+
+    Returns
+    -------
+    A tensor of shape (batch, seq_len1, seq_len2, num_perspective) consisting
+    multi-perspective matching resultsof the shape
+    """
+    num_perspective = weight.size(0)
+
+    # (1, num_perspective, 1, hidden_size)
+    weight = weight.unsqueeze(0).unsqueeze(2)
+
+    # (batch, num_perspective, seq_len*, hidden_size)
+    vector1 = weight * vector1.unsqueeze(1).expand(-1, num_perspective, -1, -1)
+    vector2 = weight * vector2.unsqueeze(1).expand(-1, num_perspective, -1, -1)
+
+    # (batch, num_perspective, seq_len*, 1)
+    v1_norm = vector1.norm(p=2, dim=3, keepdim=True)
+    v2_norm = vector2.norm(p=2, dim=3, keepdim=True)
+
+    # (batch, num_perspective, seq_len1, seq_len2)
+    mul_result = torch.matmul(vector1, vector2.transpose(2, 3))
+    norm_value = v1_norm * v2_norm.transpose(2, 3)
+
+    # (batch, seq_len1, seq_len2, num_perspective)
+    return (mul_result / norm_value.clamp(min=eps)).permute(0, 2, 3, 1)
+
+
+class BiMPMMatching(nn.Module, FromParams):
```

Comment:
We prefer to use [Google's definition of camel case](https://github.com/allenai/allennlp/blob/master/STYLE.md#naming), where this would be `BiMpmMatching` (and `BiMpm` instead of `BiMPM`).